File: david/sessions/2020/medium/102_binary_tree_level_order_traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def levelOrder(self, root: TreeNode) -> List[List[int]]:
        if not root:
            return []
        
        result = [[root.val]]
        visited = set([root])
        q = [(root, 0)]
        
        while q:
            node, level = q.pop(0)
            
            if len(result) < level+1:
                result.append([])
            
            if node not in visited:
                visited.add(node)
                result[level].append(node.val)
            
            if node.left and node.left not in visited:
                q.append((node.left, level + 1))
            if node.right and node.right not in visited:
                q.append((node.right, level + 1))
                
        return result

File: pia/2020/medium/102_binary_tree_level_order_traversal.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[][]}
 */

var levelOrder = function (root) {
  if (!root) {
    return [];
  }

  let tree = [];
  let result = [];

  tree.push(root);

  while (tree.length > 0) {
    const length = tree.length;
    let temp = [];

    for (let i = 0; i < length; i++) {
      const node = tree.shift();
      temp.push(node.val);

      if (node.left) {
        tree.push(node.left);
      }
      if (node.right) {
        tree.push(node.right);
      }
    }
    result.push(temp);
  }
  return result;
};

